{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0661641-6ae6-4d6b-bf2c-582bb80db01b",
   "metadata": {},
   "source": [
    "# Allele Length Concordance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf84388-bc3d-44fa-96af-92104bcea9bc",
   "metadata": {},
   "source": [
    "## For ATaRVa, TRGT, TREAT, Straglr & GangSTR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f934b29-72a4-43fd-8c5f-a92993b7e8a1",
   "metadata": {},
   "source": [
    "### Generating Truth set from Assembly (requires bed output from AAM.py)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c26cc0c-fb47-40a8-a645-5082bbc5ea62",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### 2.9 MILLION LOCI\n",
    "##\n",
    "# variable 'value' & '*at_motif_dict[key]' extraction index position vary based on the BED input file provided to the AAM.py\n",
    "import sys\n",
    "mat_dict = {}\n",
    "mat_motif_dict = {}\n",
    "with open('Maternal_allele.bed') as mh:\n",
    "    for line in mh:\n",
    "        line = line.strip().split('\\t')\n",
    "        key = f'{line[0]}:{line[1]}-{line[2]}'\n",
    "        value = line[-3] # allele length value\n",
    "        if value.isnumeric():\n",
    "            mat_dict[key] = int(value)\n",
    "            # mat_dict[key] = int(value)+1 # for straglr correction\n",
    "            mat_motif_dict[key] = int(line[4]) # This has to be motif length as integer\n",
    "pat_dict = {}\n",
    "pat_motif_dict = {}\n",
    "with open('Paternal_allele.bed') as ph:\n",
    "    for line in ph:\n",
    "        line = line.strip().split('\\t')\n",
    "        key = f'{line[0]}:{line[1]}-{line[2]}'\n",
    "        value = line[-3] # allele length value\n",
    "        if value.isnumeric():\n",
    "            pat_dict[key] = int(value)\n",
    "            # pat_dict[key] = int(value)+1 # for straglr correction\n",
    "            pat_motif_dict[key] = int(line[4]) # This has to be motif length as integer \n",
    "            \n",
    "print('length of Mat and Pat assembly dict = ', len(mat_dict), len(pat_dict))\n",
    "print('\\n\\n')\n",
    "\n",
    "# Merging allele lengths from the parent into a dict\n",
    "Assembly_dict = {}\n",
    "Assembly_motif_dict = {}\n",
    "for key in mat_dict:\n",
    "    if key in pat_dict:\n",
    "        if mat_motif_dict[key] == pat_motif_dict[key]:\n",
    "            Assembly_dict[key] = {mat_dict[key], pat_dict[key]}\n",
    "            Assembly_motif_dict[key] = mat_motif_dict[key]\n",
    "            \n",
    "print('length of assembly dict = ', len(Assembly_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b3f107d-a448-4fce-add4-b67708504928",
   "metadata": {},
   "source": [
    "### Categorising the truth set based on zygosity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddbbbfd5-5254-4c7a-a257-a1c2d95e93ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "Assembly_homozygous_dict = {}\n",
    "Assembly_heterozygous_dict = {}\n",
    "nonzygous = 0\n",
    "for locus_key in Assembly_dict:\n",
    "    if len(Assembly_dict[locus_key]) == 1:\n",
    "        Assembly_homozygous_dict[locus_key] = Assembly_dict[locus_key]\n",
    "    elif len(Assembly_dict[locus_key]) == 2:\n",
    "        Assembly_heterozygous_dict[locus_key] = Assembly_dict[locus_key]\n",
    "    else: nonzygous += 1\n",
    "print('length of Assembly_homozygous_dict = ', len(Assembly_homozygous_dict))\n",
    "print('length of Assembly_heterozygous_dict = ', len(Assembly_heterozygous_dict))\n",
    "print('length of nonzygous = ', nonzygous)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a82392a-7117-46f9-b38f-52d0dc284949",
   "metadata": {},
   "source": [
    "### Allele length extraction from ATaRVa VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b3a47ed-191d-4b60-a966-2f37572f4c00",
   "metadata": {},
   "outputs": [],
   "source": [
    "N_sim4_dict = {}\n",
    "with open('atarva.vcf') as sh:\n",
    "    for line in sh:\n",
    "        if line[0]=='#': continue\n",
    "        line = line.strip().split('\\t')\n",
    "        if line[6]!='PASS': continue\n",
    "        end = line[7].split(';')[3].split('=')[-1]\n",
    "        key = f'{line[0]}:{line[1]}-{end}'\n",
    "        genotype = line[9].split(':')[0] \n",
    "        gen = {int(genotype[0]), int(genotype[-1])} \n",
    "        if len(gen) == 1:\n",
    "            zygous = 0 # homo\n",
    "        else: zygous = 1 # hetero\n",
    "        allele = line[9].split(':')[1].split(',')\n",
    "        if '|' in line[9].split(':')[0]:\n",
    "            type = 1 # ambiguous\n",
    "        else: type = 0\n",
    "        value = [{int(allele[0]), int(allele[1])}, zygous, type]\n",
    "        N_sim4_dict[key] = value\n",
    "print('length of simplex_4kHz dict = ', len(N_sim4_dict))\n",
    "\n",
    "\n",
    "N_sim5_dict = {}\n",
    "print('length of simplex_5kHz dict = ', len(N_sim5_dict))\n",
    "\n",
    "\n",
    "N_simHQ_dict = {}\n",
    "print('length of simplexHQ dict = ', len(N_simHQ_dict))\n",
    "\n",
    "N_dup_dict = {}\n",
    "print('length of duplex dict = ', len(N_dup_dict))\n",
    "\n",
    "N_hifi_dict = {}\n",
    "print('length of Hifi dict = ', len(N_hifi_dict))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153286dd-8e5a-423b-aac5-bc3d0184a890",
   "metadata": {},
   "source": [
    "### Allele length extraction from TRGT VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d197db69-1bfe-4d46-9874-46a61a9b5ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "N_hifi_dict = {}\n",
    "with gzip.open(\"trgt.vcf.gz\", 'rb') as fh:\n",
    "    for line in fh:\n",
    "        line = line.decode('utf-8') # dirctly use 'rt'\n",
    "        if line[0]!=\"#\":\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            if line[9][0] == '.': continue\n",
    "            \n",
    "            info_split = line[7].split(\";\")\n",
    "            start = int(line[1])# start\n",
    "            end = info_split[1].split('=')[1] # end\n",
    "            key = f'{line[0]}:{start}-{end}'\n",
    "            alleles = line[9].split(':')[1].split(',')\n",
    "            allele_set = set()\n",
    "            for i in alleles:\n",
    "                allele_set.add(int(i))\n",
    "            type = 0\n",
    "            if len(allele_set) == 1:\n",
    "                zygous = 0\n",
    "            else: zygous = 1\n",
    "            value = [allele_set, zygous, type]\n",
    "            N_hifi_dict[key] = value\n",
    "print('length of Hifi dict = ', len(N_hifi_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37c3e29b-a294-4efd-bc49-9ce2173eac7e",
   "metadata": {},
   "source": [
    "### Allele length extraction from TREAT VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "077f62ba-ce3a-43b5-8a8a-d5bb405de72d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "N_hifi_dict = {}\n",
    "with gzip.open(\"treat.vcf.gz\", 'rb') as fh:\n",
    "    for line in fh:\n",
    "        line = line.decode('utf-8')\n",
    "        if line[0]!=\"#\":\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            sample = line[9].split(';')\n",
    "            if line[6]!='PASS': continue\n",
    "            key = line[2]\n",
    "            motif = sample[2].split('|')\n",
    "            if ('nan' in motif) or ('NA' in motif):\n",
    "                GT = sample[1].split('|')\n",
    "                if ('nan' in GT) or ('NA' in GT): continue\n",
    "                allele_set = set([int(float(i)) for i in GT])\n",
    "                type = 0\n",
    "                if len(allele_set) == 1:\n",
    "                    zygous = 0\n",
    "                else: zygous = 1\n",
    "                value = [allele_set, zygous, type]\n",
    "                N_hifi_dict[key] = value\n",
    "                continue\n",
    "\n",
    "            motif1 = motif[0].split('+')\n",
    "            motif2 = motif[1].split('+')\n",
    "            alleles = sample[3].split('|')\n",
    "            allele1 = alleles[0].split('+')\n",
    "            allele2 = alleles[1].split('+')\n",
    "            al_len1 = 0\n",
    "            al_len2 = 0\n",
    "            for i,al in enumerate(allele1):\n",
    "                al_len1 += float(al)*len(motif1[i])\n",
    "            for i,al in enumerate(allele2):\n",
    "                al_len2 += float(al)*len(motif2[i])\n",
    "            allele_set = set([round(al_len1), round(al_len2)])\n",
    "            type = 0\n",
    "            if len(allele_set) == 1:\n",
    "                zygous = 0\n",
    "            else: zygous = 1\n",
    "            value = [allele_set, zygous, type]\n",
    "            N_hifi_dict[key] = value\n",
    "            \n",
    "print('len of N_hifi_dict = ', len(N_hifi_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc9b6b00-7a3c-4e7f-998e-013d86ec0d29",
   "metadata": {},
   "source": [
    "### Allele length extraction from Straglr VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a1e41fe-a0d7-4f75-a672-da557fb2f6c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For straglr, its bed output file is used to extract Allele lengths\n",
    "# Also straglr report allele length with extra 1 bp, so the `truth set` has to be modified. Correct this by uncomment the 14, 25th line in Truth set code and comment the below line\n",
    "N_hifi_dict = {}\n",
    "with open('straglr.bed', 'r') as hih:\n",
    "    next(hih)\n",
    "    for line in hih:\n",
    "        line = line.strip().split('\\t')\n",
    "        key = f'{line[0]}:{line[1]}-{line[2]}'\n",
    "        allele_set = set()\n",
    "        if line[4]!='-':\n",
    "            allele_set.add(int(float(line[4])))\n",
    "        if line[7] != '-':\n",
    "            allele_set.add(int(float(line[7])))\n",
    "        type = 0\n",
    "        if len(allele_set) == 1:\n",
    "            zygous = 0\n",
    "        else: zygous = 1\n",
    "        value = [allele_set, zygous, type]\n",
    "        N_hifi_dict[key] = value\n",
    "print('length of hifi dict = ', len(N_hifi_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dba19abe-de5c-4735-8c32-1f1c65f2c1db",
   "metadata": {},
   "source": [
    "### Allele length extraction from GangSTR VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8bef7f3-bed2-442a-bf44-06284f6541fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For gangstr also do the 'straglr correction' in the `truth set` script\n",
    "N_hifi_dict = {}\n",
    "import gzip\n",
    "with gzip.open('Gangstr.vcf.gz', 'rb') as srh:\n",
    "    for line in srh:\n",
    "        line = line.decode('utf-8')\n",
    "        if line[0]=='#': continue\n",
    "        line = line.strip().split(\"\\t\")\n",
    "        if line[9][0] == '.': continue\n",
    "        info_split = line[7].split(\";\")\n",
    "        start = int(line[1])\n",
    "        end = int(info_split[0].split('=')[1])\n",
    "        motif = int(info_split[2].split('=')[1])\n",
    "        key = f'{line[0]}:{start}-{end}'\n",
    "        alleles = [int(i)*motif for i in line[9].split(':')[3].split(',')]\n",
    "        allele_set = set(alleles)\n",
    "        type = 0\n",
    "        if len(allele_set) == 1:\n",
    "            zygous = 0\n",
    "        else: zygous = 1\n",
    "        value = [allele_set, zygous, type]\n",
    "\n",
    "        N_hifi_dict[key] = value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0249b6cd-7f87-41a2-b9e8-b5b3ad26635e",
   "metadata": {},
   "source": [
    "## Concordance Calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f9e29b5-f6c1-408c-9428-1008b1974029",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx,dicts in enumerate([N_sim4_dict, N_sim5_dict, N_simHQ_dict, N_dup_dict, N_hifi_dict]):\n",
    "    if len(dicts) == 0: continue\n",
    "    if idx == 0: print('***Simplex4kHz***')\n",
    "    elif idx == 1: print('***Simplex5kHz***')\n",
    "    elif idx == 2: print('***Simplex-HQ***')\n",
    "    elif idx == 3: print('***Duplex***')\n",
    "    else: print('***Hifi***')\n",
    "    common_loci_btw_assembly_and_sim = len(set(list(Assembly_dict.keys())) & set(list(dicts.keys())))\n",
    "    print('common_loci_btw_assembly_and_sample = ', common_loci_btw_assembly_and_sim)\n",
    "\n",
    "    common_loci_btw_HOMO_assembly_and_sim = len(set(list(Assembly_homozygous_dict.keys())) & set(list(dicts.keys())))\n",
    "    print('common_loci_btw_HOMO_assembly_and_sample = ', common_loci_btw_HOMO_assembly_and_sim)\n",
    "\n",
    "    common_loci_btw_HETERO_assembly_and_sim = len(set(list(Assembly_heterozygous_dict.keys())) & set(list(dicts.keys())))\n",
    "    print('common_loci_btw_HETERO_assembly_and_sample = ', common_loci_btw_HETERO_assembly_and_sim)\n",
    "\n",
    "    # both tool & assembly says as Homozygous\n",
    "    sim_homo_homo_match = 0\n",
    "    sim_homo_homo_1bp_diff = 0\n",
    "    sim_homo_homo_40m_diff = 0 # allele length differ by 40% of motif length \n",
    "    sim_homo_homo_mismatch = 0 # difference beyond 40%\n",
    "\n",
    "    # Assembly is Homozygous but tool says Heterozygous\n",
    "    sim_homo_hetero_match_diff1bp = 0 # one allele is matching another one os differing by 1 bp\n",
    "    sim_homo_hetero_match_40m_diff = 0\n",
    "    sim_homo_hetero_match_mismatch = 0\n",
    "    sim_homo_hetero_both_1bpdiff = 0\n",
    "    sim_homo_hetero_both_40m_diff = 0\n",
    "    sim_homo_hetero_1bdfiff_40m_diff = 0\n",
    "    sim_homo_hetero_1bpdiff_mismatch = 0\n",
    "    sim_homo_hetero_40m_diff_mismatch = 0\n",
    "    sim_homo_hetero_both_mismatch = 0\n",
    "\n",
    "\n",
    "    for locus_key in Assembly_homozygous_dict:\n",
    "        \n",
    "        if locus_key in dicts:\n",
    "            Value = dicts[locus_key]\n",
    "            motif_thresh = 0.4*Assembly_motif_dict[locus_key]\n",
    "\n",
    "            if Value[1] == 0: # homozygous\n",
    "                sim_allele = list(Value[0])[0]\n",
    "                ass_allele = list(Assembly_homozygous_dict[locus_key])[0]\n",
    "                if sim_allele == ass_allele:\n",
    "\n",
    "                    sim_homo_homo_match +=1\n",
    "\n",
    "                        \n",
    "                elif abs(sim_allele-ass_allele) == 1:\n",
    "\n",
    "                    sim_homo_homo_1bp_diff += 1\n",
    "                elif abs(sim_allele-ass_allele) <= motif_thresh: sim_homo_homo_40m_diff += 1\n",
    "                else:\n",
    "                    sim_homo_homo_mismatch += 1\n",
    "\n",
    "            else:\n",
    "                sim_allele = Value[0]\n",
    "                ass_allele = Assembly_homozygous_dict[locus_key]\n",
    "                if len(sim_allele-ass_allele) == 1:\n",
    "                    if abs(list(sim_allele-ass_allele)[0] - list(ass_allele)[0]) == 1: \n",
    "                        sim_homo_hetero_match_diff1bp += 1\n",
    "                    elif abs(list(sim_allele-ass_allele)[0] - list(ass_allele)[0]) <= motif_thresh:\n",
    "                        sim_homo_hetero_match_40m_diff += 1\n",
    "                    else: \n",
    "                        sim_homo_hetero_match_mismatch += 1\n",
    "                else:\n",
    "                    if (abs(list(sim_allele)[0] - list(ass_allele)[0]) == 1) and (abs(list(sim_allele)[1] - list(ass_allele)[0]) == 1):\n",
    "                        sim_homo_hetero_both_1bpdiff += 1\n",
    "                    elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) == 1) or (abs(list(sim_allele)[1] - list(ass_allele)[0]) == 1):\n",
    "                        list_sim_allele = list(sim_allele)\n",
    "                        if abs(list_sim_allele[0] - list(ass_allele)[0]) == 1:\n",
    "                            if abs(list(sim_allele)[1] - list(ass_allele)[0]) <= motif_thresh:\n",
    "                                sim_homo_hetero_1bdfiff_40m_diff +=1\n",
    "                            else:\n",
    "                                sim_homo_hetero_1bpdiff_mismatch += 1\n",
    "                        else:\n",
    "                            if abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh:\n",
    "                                sim_homo_hetero_1bdfiff_40m_diff +=1\n",
    "                            else:\n",
    "                                sim_homo_hetero_1bpdiff_mismatch += 1\n",
    "                    elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh) and (abs(list(sim_allele)[1] - list(ass_allele)[0]) <= motif_thresh):\n",
    "                        sim_homo_hetero_both_40m_diff += 1    \n",
    "                    elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh) or (abs(list(sim_allele)[1] - list(ass_allele)[0]) <= motif_thresh):    \n",
    "                        sim_homo_hetero_40m_diff_mismatch += 1\n",
    "                    else: sim_homo_hetero_both_mismatch += 1\n",
    "\n",
    "    print(f\"\"\"\\n    \n",
    "    homo_homo_match = {sim_homo_homo_match}\n",
    "    homo_homo_1bp_diff = {sim_homo_homo_1bp_diff}\n",
    "    homo_homo_40m_diff = {sim_homo_homo_40m_diff}\n",
    "    homo_homo_mismatch = {sim_homo_homo_mismatch}\n",
    "    \n",
    "    homo_hetero_match_diff1bp = {sim_homo_hetero_match_diff1bp}\n",
    "    homo_hetero_both_1bpdiff = {sim_homo_hetero_both_1bpdiff}\n",
    "    homo_hetero_match_40m_diff = {sim_homo_hetero_match_40m_diff}\n",
    "    homo_hetero_both_40m_diff = {sim_homo_hetero_both_40m_diff}\n",
    "    homo_hetero_match_mismatch = {sim_homo_hetero_match_mismatch}\n",
    "    homo_hetero_1bdfiff_40m_diff = {sim_homo_hetero_1bdfiff_40m_diff}\n",
    "    homo_hetero_1bpdiff_mismatch = {sim_homo_hetero_1bpdiff_mismatch}\n",
    "    homo_hetero_40m_diff_mismatch = {sim_homo_hetero_40m_diff_mismatch}\n",
    "    homo_hetero_both_mismatch = {sim_homo_hetero_both_mismatch}\"\"\")\n",
    "    \n",
    "    print('\\nHomozygous concordance = ', ((sim_homo_homo_match+sim_homo_homo_1bp_diff+sim_homo_hetero_match_diff1bp+sim_homo_hetero_both_1bpdiff)/common_loci_btw_HOMO_assembly_and_sim)*100)\n",
    "    # the above categories are considered for final calculation of concordance\n",
    "\n",
    "\n",
    "    \n",
    "    sim_hetero_homo_one_match = 0\n",
    "    sim_hetero_homo_1bp_diff = 0\n",
    "    sim_hetero_homo_40m_diff = 0\n",
    "    sim_hetero_homo_mismatch = 0\n",
    "    \n",
    "    sim_hetero_hetero_match_ = 0\n",
    "    sim_hetero_hetero_match_diff1bp = 0\n",
    "    sim_hetero_hetero_match_mismatch = 0\n",
    "    sim_hetero_hetero_both_1bpdiff = 0\n",
    "    sim_hetero_hetero_1bpdiff_mismatch = 0\n",
    "    sim_hetero_hetero_both_mismatch = 0\n",
    "    sim_hetero_hetero_match_40m_diff = 0\n",
    "    sim_hetero_hetero_1bpdiff_40m_diff = 0\n",
    "    sim_hetero_hetero_both_40m_diff = 0\n",
    "    sim_hetero_hetero_40m_diff_mismatch = 0\n",
    "\n",
    "    for locus_key in Assembly_heterozygous_dict:\n",
    "    \n",
    "        if locus_key in dicts:\n",
    "            Value = dicts[locus_key]\n",
    "            motif_thresh = 0.4*Assembly_motif_dict[locus_key]\n",
    "\n",
    "            if Value[1] == 0: # homozygous\n",
    "                \n",
    "                sim_allele = Value[0]\n",
    "                ass_allele = Assembly_heterozygous_dict[locus_key]\n",
    "                if len(sim_allele & ass_allele) == 1:\n",
    "                    sim_hetero_homo_one_match += 1\n",
    "                        \n",
    "                elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) == 1) or (abs(list(sim_allele)[0] - list(ass_allele)[1]) == 1):\n",
    "                    sim_hetero_homo_1bp_diff += 1\n",
    "                            \n",
    "                elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh) or (abs(list(sim_allele)[0] - list(ass_allele)[1]) <= motif_thresh):\n",
    "                    sim_hetero_homo_40m_diff += 1\n",
    "\n",
    "                else:\n",
    "                    sim_hetero_homo_mismatch += 1\n",
    "                \n",
    "            \n",
    "            else:\n",
    "                sim_allele = Value[0]\n",
    "                ass_allele = Assembly_heterozygous_dict[locus_key]\n",
    "                if sim_allele == ass_allele: \n",
    "                    sim_hetero_hetero_match_ += 1\n",
    "                elif len(sim_allele & ass_allele) == 1:\n",
    "                    diff = list(sim_allele ^ ass_allele)\n",
    "                    if abs(diff[0] - diff[1]) == 1:\n",
    "                       sim_hetero_hetero_match_diff1bp += 1\n",
    "                    elif abs(diff[0] - diff[1]) <= motif_thresh:\n",
    "                        sim_hetero_hetero_match_40m_diff += 1\n",
    "                    else: sim_hetero_hetero_match_mismatch += 1\n",
    "                else:\n",
    "                    list_ass_allele = list(ass_allele)\n",
    "                    list_sample_allele = list(sim_allele)\n",
    "                    if ((abs(list_ass_allele[0] - list_sample_allele[0]) == 1) and (abs(list_ass_allele[1] - list_sample_allele[1]) == 1)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) == 1) and (abs(list_ass_allele[1] - list_sample_allele[0]) == 1)):\n",
    "                        sim_hetero_hetero_both_1bpdiff += 1\n",
    "                    elif ((abs(list_ass_allele[0] - list_sample_allele[0]) == 1) or (abs(list_ass_allele[1] - list_sample_allele[1]) == 1)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) == 1) or (abs(list_ass_allele[1] - list_sample_allele[0]) == 1)):\n",
    "                        if (abs(list_ass_allele[0] - list_sample_allele[0]) == 1) or (abs(list_ass_allele[1] - list_sample_allele[1]) == 1):\n",
    "                            if abs(list_ass_allele[0] - list_sample_allele[0]) == 1:\n",
    "                                if abs(list_ass_allele[1] - list_sample_allele[1]) <= motif_thresh:\n",
    "                                    sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                                else:\n",
    "                                    sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                            elif abs(list_ass_allele[0] - list_sample_allele[0]) <= motif_thresh:\n",
    "                                sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                            else:\n",
    "                                sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                        elif abs(list_ass_allele[0] - list_sample_allele[1]) == 1:\n",
    "                            if abs(list_ass_allele[1] - list_sample_allele[0]) <= motif_thresh:\n",
    "                                sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                            else:\n",
    "                                sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                        elif abs(list_ass_allele[1] - list_sample_allele[0]) == 1:\n",
    "                            if abs(list_ass_allele[0] - list_sample_allele[1]) <= motif_thresh:\n",
    "                                sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                            else:\n",
    "                                sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                        else:\n",
    "                            sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                            \n",
    "                    elif ((abs(list_ass_allele[0] - list_sample_allele[0]) <= motif_thresh) and (abs(list_ass_allele[1] - list_sample_allele[1]) <= motif_thresh)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) <= motif_thresh) and (abs(list_ass_allele[1] - list_sample_allele[0]) <= motif_thresh)):\n",
    "                        sim_hetero_hetero_both_40m_diff += 1\n",
    "                    elif ((abs(list_ass_allele[0] - list_sample_allele[0]) <= motif_thresh) or (abs(list_ass_allele[1] - list_sample_allele[1]) <= motif_thresh)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) <= motif_thresh) or (abs(list_ass_allele[1] - list_sample_allele[0]) <= motif_thresh)):\n",
    "                        sim_hetero_hetero_40m_diff_mismatch += 1\n",
    "                    else: sim_hetero_hetero_both_mismatch += 1\n",
    "\n",
    "\n",
    "    print(f'''\\n    \n",
    "    hetero_homo_one_match = {sim_hetero_homo_one_match}\n",
    "    hetero_homo_1bp_diff = {sim_hetero_homo_1bp_diff}\n",
    "    hetero_homo_40m_diff = {sim_hetero_homo_40m_diff}\n",
    "    hetero_homo_mismatch = {sim_hetero_homo_mismatch}\n",
    "    \n",
    "    hetero_hetero_match_ = {sim_hetero_hetero_match_}\n",
    "    hetero_hetero_match_diff1bp = {sim_hetero_hetero_match_diff1bp}\n",
    "    hetero_hetero_match_40m_diff = {sim_hetero_hetero_match_40m_diff}\n",
    "    hetero_hetero_match_mismatch = {sim_hetero_hetero_match_mismatch}\n",
    "    hetero_hetero_both_1bpdiff = {sim_hetero_hetero_both_1bpdiff}\n",
    "    hetero_hetero_1bpdiff_40m_diff = {sim_hetero_hetero_1bpdiff_40m_diff}\n",
    "    hetero_hetero_1bpdiff_mismatch = {sim_hetero_hetero_1bpdiff_mismatch}\n",
    "    hetero_hetero_both_40m_diff = {sim_hetero_hetero_both_40m_diff}\n",
    "    hetero_hetero_40m_diff_mismatch = {sim_hetero_hetero_40m_diff_mismatch}\n",
    "    hetero_hetero_both_mismatch = {sim_hetero_hetero_both_mismatch}\\n\\n\\n''')\n",
    "    \n",
    "    print('\\nHeterozygous concordance = ', ((sim_hetero_hetero_match_+sim_hetero_hetero_match_diff1bp+sim_hetero_hetero_both_1bpdiff)/common_loci_btw_HETERO_assembly_and_sim)*100)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
