{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bedc0954-e530-468d-b792-68831263790d",
   "metadata": {},
   "source": [
    "### LongTR report allele length difference rather than the absolute length"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bdd5236-a295-4e14-b061-2d028269c32b",
   "metadata": {},
   "source": [
    "### Generating Truth set from Assembly (requires bed output from AAM.py)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb1b57af-c982-4185-9738-da2c14783c43",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### 2.9 MILLION LOCI\n",
    "##\n",
    "# variable 'value' & '*at_motif_dict[key]' extraction index position vary based on the BED input file provided to the AAM.py\n",
    "import sys\n",
    "mat_dict = {}\n",
    "mat_motif_dict = {}\n",
    "with open('Maternal_allele.bed') as mh:\n",
    "    for line in mh:\n",
    "        line = line.strip().split('\\t')\n",
    "        key = f'{line[0]}:{line[1]}-{line[2]}'\n",
    "        value = line[-3] # allele length value\n",
    "        if value.isnumeric():\n",
    "            mat_dict[key] = int(value) - (int(line[2]) - int(line[1])) # calc length difference wrt reference allele length\n",
    "            mat_motif_dict[key] = int(float(line[4])) # This has to be motif length as integer\n",
    "pat_dict = {}\n",
    "pat_motif_dict = {}\n",
    "with open('Paternal_allele.bed') as ph:\n",
    "    for line in ph:\n",
    "        line = line.strip().split('\\t')\n",
    "        key = f'{line[0]}:{line[1]}-{line[2]}'\n",
    "        value = line[-3] # allele length value\n",
    "        if value.isnumeric():\n",
    "            pat_dict[key] = int(value) - (int(line[2]) - int(line[1])) # calc length difference wrt reference allele length\n",
    "            pat_motif_dict[key] = int(float(line[4])) # This has to be motif length as integer \n",
    "            \n",
    "print('length of Mat and Pat assembly dict = ', len(mat_dict), len(pat_dict))\n",
    "print('\\n\\n')\n",
    "\n",
    "# Merging allele lengths from the parent into a dict\n",
    "Assembly_dict = {}\n",
    "Assembly_motif_dict = {}\n",
    "for key in mat_dict:\n",
    "    if key in pat_dict:\n",
    "        if mat_motif_dict[key] == pat_motif_dict[key]:\n",
    "            Assembly_dict[key] = {mat_dict[key], pat_dict[key]}\n",
    "            Assembly_motif_dict[key] = mat_motif_dict[key]\n",
    "            \n",
    "print('length of assembly dict = ', len(Assembly_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b99ee5b6-0447-46d0-82ec-b75c231544f6",
   "metadata": {},
   "source": [
    "### Categorising the truth set based on zygosity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fac6e352-9c52-4dc9-a04a-b5a4fcc38272",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "length of Assembly_homozygous_dict =  2575955\n",
      "length of Assembly_heterozygous_dict =  199015\n",
      "length of nonzygous =  0\n"
     ]
    }
   ],
   "source": [
    "Assembly_homozygous_dict = {}\n",
    "Assembly_heterozygous_dict = {}\n",
    "nonzygous = 0\n",
    "for locus_key in Assembly_dict:\n",
    "    if len(Assembly_dict[locus_key]) == 1:\n",
    "        Assembly_homozygous_dict[locus_key] = Assembly_dict[locus_key]\n",
    "    elif len(Assembly_dict[locus_key]) == 2:\n",
    "        Assembly_heterozygous_dict[locus_key] = Assembly_dict[locus_key]\n",
    "    else: nonzygous += 1\n",
    "print('length of Assembly_homozygous_dict = ', len(Assembly_homozygous_dict))\n",
    "print('length of Assembly_heterozygous_dict = ', len(Assembly_heterozygous_dict))\n",
    "print('length of nonzygous = ', nonzygous)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b437280b-ecc2-44c9-a9a5-bd5a3f6b0ca1",
   "metadata": {},
   "source": [
    "### Allele length extraction from LongTR VCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16ee73a2-306e-41c1-b75e-39b55ecd4b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "N_hifi_dict = {}\n",
    "with gzip.open(\"tr_calls.vcf.gz\", 'rb') as fh:\n",
    "    for line in fh:\n",
    "        line = line.decode('utf-8') # use 'rb'\n",
    "        if line[0]!=\"#\":\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            if line[9]=='.': continue\n",
    "            info_split = line[7].split(\";\")\n",
    "            start = info_split[0].split('=')[1] # start\n",
    "            end = info_split[1].split('=')[1] # end\n",
    "            key = f'{line[0]}:{start}-{end}'\n",
    "            alleles = line[9].split(':')[1].split('|')\n",
    "            allele_set = set()\n",
    "            for i in alleles:\n",
    "                allele_set.add(int(i))\n",
    "            type = 0\n",
    "            if len(allele_set) == 1:\n",
    "                zygous = 0\n",
    "            else: zygous = 1\n",
    "            value = [allele_set, zygous, type]\n",
    "\n",
    "            N_hifi_dict[key] = value\n",
    "print('length of HiFi dict = ', len(N_hifi_dict))\n",
    "\n",
    "N_sim5_dict = {}\n",
    "print('length of simplex_5kHz dict = ', len(N_sim5_dict))\n",
    "\n",
    "\n",
    "N_simHQ_dict = {}\n",
    "print('length of simplexHQ dict = ', len(N_simHQ_dict))\n",
    "\n",
    "N_dup_dict = {}\n",
    "print('length of duplex dict = ', len(N_dup_dict))\n",
    "\n",
    "N_sim4_dict = {}\n",
    "print('length of simplex_4kHz dict = ', len(N_sim4_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27a9f7b1-cae3-4fdb-9aae-fd140fd41ebb",
   "metadata": {},
   "source": [
    "## Concordance Calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b981e04-e2bd-46e6-928a-43fefcb5f3f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx,dicts in enumerate([N_sim4_dict, N_sim5_dict, N_simHQ_dict, N_dup_dict, N_hifi_dict]):\n",
    "    if len(dicts) == 0: continue\n",
    "    if idx == 0: print('***Simplex4kHz***')\n",
    "    elif idx == 1: print('***Simplex5kHz***')\n",
    "    elif idx == 2: print('***Simplex-HQ***')\n",
    "    elif idx == 3: print('***Duplex***')\n",
    "    else: print('***Hifi***')\n",
    "    common_loci_btw_assembly_and_sim = len(set(list(Assembly_dict.keys())) & set(list(dicts.keys())))\n",
    "    print('common_loci_btw_assembly_and_sample = ', common_loci_btw_assembly_and_sim)\n",
    "\n",
    "    common_loci_btw_HOMO_assembly_and_sim = len(set(list(Assembly_homozygous_dict.keys())) & set(list(dicts.keys())))\n",
    "    print('common_loci_btw_HOMO_assembly_and_sample = ', common_loci_btw_HOMO_assembly_and_sim)\n",
    "\n",
    "    common_loci_btw_HETERO_assembly_and_sim = len(set(list(Assembly_heterozygous_dict.keys())) & set(list(dicts.keys())))\n",
    "    print('common_loci_btw_HETERO_assembly_and_sample = ', common_loci_btw_HETERO_assembly_and_sim)\n",
    "\n",
    "    # both tool & assembly says as Homozygous\n",
    "    sim_homo_homo_match = 0\n",
    "    sim_homo_homo_1bp_diff = 0\n",
    "    sim_homo_homo_40m_diff = 0 # allele length differ by 40% of motif length \n",
    "    sim_homo_homo_mismatch = 0 # difference beyond 40%\n",
    "\n",
    "    # Assembly is Homozygous but tool says Heterozygous\n",
    "    sim_homo_hetero_match_diff1bp = 0 # one allele is matching another one os differing by 1 bp\n",
    "    sim_homo_hetero_match_40m_diff = 0\n",
    "    sim_homo_hetero_match_mismatch = 0\n",
    "    sim_homo_hetero_both_1bpdiff = 0\n",
    "    sim_homo_hetero_both_40m_diff = 0\n",
    "    sim_homo_hetero_1bdfiff_40m_diff = 0\n",
    "    sim_homo_hetero_1bpdiff_mismatch = 0\n",
    "    sim_homo_hetero_40m_diff_mismatch = 0\n",
    "    sim_homo_hetero_both_mismatch = 0\n",
    "\n",
    "\n",
    "    for locus_key in Assembly_homozygous_dict:\n",
    "        \n",
    "        if locus_key in dicts:\n",
    "            Value = dicts[locus_key]\n",
    "            motif_thresh = 0.4*Assembly_motif_dict[locus_key]\n",
    "\n",
    "            if Value[1] == 0: # homozygous\n",
    "                sim_allele = list(Value[0])[0]\n",
    "                ass_allele = list(Assembly_homozygous_dict[locus_key])[0]\n",
    "                if sim_allele == ass_allele:\n",
    "\n",
    "                    sim_homo_homo_match +=1\n",
    "\n",
    "                        \n",
    "                elif abs(sim_allele-ass_allele) == 1:\n",
    "\n",
    "                    sim_homo_homo_1bp_diff += 1\n",
    "                elif abs(sim_allele-ass_allele) <= motif_thresh: sim_homo_homo_40m_diff += 1\n",
    "                else:\n",
    "                    sim_homo_homo_mismatch += 1\n",
    "\n",
    "            else:\n",
    "                sim_allele = Value[0]\n",
    "                ass_allele = Assembly_homozygous_dict[locus_key]\n",
    "                if len(sim_allele-ass_allele) == 1:\n",
    "                    if abs(list(sim_allele-ass_allele)[0] - list(ass_allele)[0]) == 1: \n",
    "                        sim_homo_hetero_match_diff1bp += 1\n",
    "                    elif abs(list(sim_allele-ass_allele)[0] - list(ass_allele)[0]) <= motif_thresh:\n",
    "                        sim_homo_hetero_match_40m_diff += 1\n",
    "                    else: \n",
    "                        sim_homo_hetero_match_mismatch += 1\n",
    "                else:\n",
    "                    if (abs(list(sim_allele)[0] - list(ass_allele)[0]) == 1) and (abs(list(sim_allele)[1] - list(ass_allele)[0]) == 1):\n",
    "                        sim_homo_hetero_both_1bpdiff += 1\n",
    "                    elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) == 1) or (abs(list(sim_allele)[1] - list(ass_allele)[0]) == 1):\n",
    "                        list_sim_allele = list(sim_allele)\n",
    "                        if abs(list_sim_allele[0] - list(ass_allele)[0]) == 1:\n",
    "                            if abs(list(sim_allele)[1] - list(ass_allele)[0]) <= motif_thresh:\n",
    "                                sim_homo_hetero_1bdfiff_40m_diff +=1\n",
    "                            else:\n",
    "                                sim_homo_hetero_1bpdiff_mismatch += 1\n",
    "                        else:\n",
    "                            if abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh:\n",
    "                                sim_homo_hetero_1bdfiff_40m_diff +=1\n",
    "                            else:\n",
    "                                sim_homo_hetero_1bpdiff_mismatch += 1\n",
    "                    elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh) and (abs(list(sim_allele)[1] - list(ass_allele)[0]) <= motif_thresh):\n",
    "                        sim_homo_hetero_both_40m_diff += 1    \n",
    "                    elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh) or (abs(list(sim_allele)[1] - list(ass_allele)[0]) <= motif_thresh):    \n",
    "                        sim_homo_hetero_40m_diff_mismatch += 1\n",
    "                    else: sim_homo_hetero_both_mismatch += 1\n",
    "\n",
    "    print(f\"\"\"\\n    \n",
    "    homo_homo_match = {sim_homo_homo_match}\n",
    "    homo_homo_1bp_diff = {sim_homo_homo_1bp_diff}\n",
    "    homo_homo_40m_diff = {sim_homo_homo_40m_diff}\n",
    "    homo_homo_mismatch = {sim_homo_homo_mismatch}\n",
    "    \n",
    "    homo_hetero_match_diff1bp = {sim_homo_hetero_match_diff1bp}\n",
    "    homo_hetero_both_1bpdiff = {sim_homo_hetero_both_1bpdiff}\n",
    "    homo_hetero_match_40m_diff = {sim_homo_hetero_match_40m_diff}\n",
    "    homo_hetero_both_40m_diff = {sim_homo_hetero_both_40m_diff}\n",
    "    homo_hetero_match_mismatch = {sim_homo_hetero_match_mismatch}\n",
    "    homo_hetero_1bdfiff_40m_diff = {sim_homo_hetero_1bdfiff_40m_diff}\n",
    "    homo_hetero_1bpdiff_mismatch = {sim_homo_hetero_1bpdiff_mismatch}\n",
    "    homo_hetero_40m_diff_mismatch = {sim_homo_hetero_40m_diff_mismatch}\n",
    "    homo_hetero_both_mismatch = {sim_homo_hetero_both_mismatch}\"\"\")\n",
    "    \n",
    "    print('\\nHomozygous concordance = ', ((sim_homo_homo_match+sim_homo_homo_1bp_diff+sim_homo_hetero_match_diff1bp+sim_homo_hetero_both_1bpdiff)/common_loci_btw_HOMO_assembly_and_sim)*100)\n",
    "    # the above categories are considered for final calculation of concordance\n",
    "\n",
    "\n",
    "    \n",
    "    sim_hetero_homo_one_match = 0\n",
    "    sim_hetero_homo_1bp_diff = 0\n",
    "    sim_hetero_homo_40m_diff = 0\n",
    "    sim_hetero_homo_mismatch = 0\n",
    "    \n",
    "    sim_hetero_hetero_match_ = 0\n",
    "    sim_hetero_hetero_match_diff1bp = 0\n",
    "    sim_hetero_hetero_match_mismatch = 0\n",
    "    sim_hetero_hetero_both_1bpdiff = 0\n",
    "    sim_hetero_hetero_1bpdiff_mismatch = 0\n",
    "    sim_hetero_hetero_both_mismatch = 0\n",
    "    sim_hetero_hetero_match_40m_diff = 0\n",
    "    sim_hetero_hetero_1bpdiff_40m_diff = 0\n",
    "    sim_hetero_hetero_both_40m_diff = 0\n",
    "    sim_hetero_hetero_40m_diff_mismatch = 0\n",
    "\n",
    "    for locus_key in Assembly_heterozygous_dict:\n",
    "    \n",
    "        if locus_key in dicts:\n",
    "            Value = dicts[locus_key]\n",
    "            motif_thresh = 0.4*Assembly_motif_dict[locus_key]\n",
    "\n",
    "            if Value[1] == 0: # homozygous\n",
    "                \n",
    "                sim_allele = Value[0]\n",
    "                ass_allele = Assembly_heterozygous_dict[locus_key]\n",
    "                if len(sim_allele & ass_allele) == 1:\n",
    "                    sim_hetero_homo_one_match += 1\n",
    "                        \n",
    "                elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) == 1) or (abs(list(sim_allele)[0] - list(ass_allele)[1]) == 1):\n",
    "                    sim_hetero_homo_1bp_diff += 1\n",
    "                            \n",
    "                elif (abs(list(sim_allele)[0] - list(ass_allele)[0]) <= motif_thresh) or (abs(list(sim_allele)[0] - list(ass_allele)[1]) <= motif_thresh):\n",
    "                    sim_hetero_homo_40m_diff += 1\n",
    "\n",
    "                else:\n",
    "                    sim_hetero_homo_mismatch += 1\n",
    "                \n",
    "            \n",
    "            else:\n",
    "                sim_allele = Value[0]\n",
    "                ass_allele = Assembly_heterozygous_dict[locus_key]\n",
    "                if sim_allele == ass_allele: \n",
    "                    sim_hetero_hetero_match_ += 1\n",
    "                elif len(sim_allele & ass_allele) == 1:\n",
    "                    diff = list(sim_allele ^ ass_allele)\n",
    "                    if abs(diff[0] - diff[1]) == 1:\n",
    "                       sim_hetero_hetero_match_diff1bp += 1\n",
    "                    elif abs(diff[0] - diff[1]) <= motif_thresh:\n",
    "                        sim_hetero_hetero_match_40m_diff += 1\n",
    "                    else: sim_hetero_hetero_match_mismatch += 1\n",
    "                else:\n",
    "                    list_ass_allele = list(ass_allele)\n",
    "                    list_sample_allele = list(sim_allele)\n",
    "                    if ((abs(list_ass_allele[0] - list_sample_allele[0]) == 1) and (abs(list_ass_allele[1] - list_sample_allele[1]) == 1)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) == 1) and (abs(list_ass_allele[1] - list_sample_allele[0]) == 1)):\n",
    "                        sim_hetero_hetero_both_1bpdiff += 1\n",
    "                    elif ((abs(list_ass_allele[0] - list_sample_allele[0]) == 1) or (abs(list_ass_allele[1] - list_sample_allele[1]) == 1)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) == 1) or (abs(list_ass_allele[1] - list_sample_allele[0]) == 1)):\n",
    "                        if (abs(list_ass_allele[0] - list_sample_allele[0]) == 1) or (abs(list_ass_allele[1] - list_sample_allele[1]) == 1):\n",
    "                            if abs(list_ass_allele[0] - list_sample_allele[0]) == 1:\n",
    "                                if abs(list_ass_allele[1] - list_sample_allele[1]) <= motif_thresh:\n",
    "                                    sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                                else:\n",
    "                                    sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                            elif abs(list_ass_allele[0] - list_sample_allele[0]) <= motif_thresh:\n",
    "                                sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                            else:\n",
    "                                sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                        elif abs(list_ass_allele[0] - list_sample_allele[1]) == 1:\n",
    "                            if abs(list_ass_allele[1] - list_sample_allele[0]) <= motif_thresh:\n",
    "                                sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                            else:\n",
    "                                sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                        elif abs(list_ass_allele[1] - list_sample_allele[0]) == 1:\n",
    "                            if abs(list_ass_allele[0] - list_sample_allele[1]) <= motif_thresh:\n",
    "                                sim_hetero_hetero_1bpdiff_40m_diff += 1\n",
    "                            else:\n",
    "                                sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                        else:\n",
    "                            sim_hetero_hetero_1bpdiff_mismatch += 1\n",
    "                            \n",
    "                    elif ((abs(list_ass_allele[0] - list_sample_allele[0]) <= motif_thresh) and (abs(list_ass_allele[1] - list_sample_allele[1]) <= motif_thresh)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) <= motif_thresh) and (abs(list_ass_allele[1] - list_sample_allele[0]) <= motif_thresh)):\n",
    "                        sim_hetero_hetero_both_40m_diff += 1\n",
    "                    elif ((abs(list_ass_allele[0] - list_sample_allele[0]) <= motif_thresh) or (abs(list_ass_allele[1] - list_sample_allele[1]) <= motif_thresh)) or ((abs(list_ass_allele[0] - list_sample_allele[1]) <= motif_thresh) or (abs(list_ass_allele[1] - list_sample_allele[0]) <= motif_thresh)):\n",
    "                        sim_hetero_hetero_40m_diff_mismatch += 1\n",
    "                    else: sim_hetero_hetero_both_mismatch += 1\n",
    "\n",
    "\n",
    "    print(f'''\\n    \n",
    "    hetero_homo_one_match = {sim_hetero_homo_one_match}\n",
    "    hetero_homo_1bp_diff = {sim_hetero_homo_1bp_diff}\n",
    "    hetero_homo_40m_diff = {sim_hetero_homo_40m_diff}\n",
    "    hetero_homo_mismatch = {sim_hetero_homo_mismatch}\n",
    "    \n",
    "    hetero_hetero_match_ = {sim_hetero_hetero_match_}\n",
    "    hetero_hetero_match_diff1bp = {sim_hetero_hetero_match_diff1bp}\n",
    "    hetero_hetero_match_40m_diff = {sim_hetero_hetero_match_40m_diff}\n",
    "    hetero_hetero_match_mismatch = {sim_hetero_hetero_match_mismatch}\n",
    "    hetero_hetero_both_1bpdiff = {sim_hetero_hetero_both_1bpdiff}\n",
    "    hetero_hetero_1bpdiff_40m_diff = {sim_hetero_hetero_1bpdiff_40m_diff}\n",
    "    hetero_hetero_1bpdiff_mismatch = {sim_hetero_hetero_1bpdiff_mismatch}\n",
    "    hetero_hetero_both_40m_diff = {sim_hetero_hetero_both_40m_diff}\n",
    "    hetero_hetero_40m_diff_mismatch = {sim_hetero_hetero_40m_diff_mismatch}\n",
    "    hetero_hetero_both_mismatch = {sim_hetero_hetero_both_mismatch}\\n\\n\\n''')\n",
    "    \n",
    "    print('\\nHeterozygous concordance = ', ((sim_hetero_hetero_match_+sim_hetero_hetero_match_diff1bp+sim_hetero_hetero_both_1bpdiff)/common_loci_btw_HETERO_assembly_and_sim)*100)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
