# MIRCHI - Mendelian Inheritance Repeat Concordance in Human Individuals

<p align=center>
  <img src="logo_Mirchi.png" alt="Logo of Mirchi" width="200"/>
</p>

MIRCHI is a Python-based tool for calculating Mendelian inheritance concordance in trio datasets of tandem repeats. It requires the pysam package for operation.

MIRCHI supports two modes of analysis:

- *All-loci mode* – Calculates Mendelian concordance for all loci specified in the input BED file.

- *Non-reference mode* – Calculates concordance only for loci where at least one family member carries a non-reference allele.

## Usage
```bash
$ python MIRCHI.py -h
```
which gives the following output:

```
usage: MIRCHI [-h] -k <FILE> -d <FILE> -m <FILE> -r <FILE> [--non_ref] [-o <FILE>] [--contigs CONTIGS [CONTIGS ...]]

Required arguments:
  -k <FILE>, --kid <FILE>
                        offspring VCF file.
  -d <FILE>, --dad <FILE>
                        paternal VCF file.
  -m <FILE>, --mom <FILE>
                        maternal VCF file.
  -r <FILE>, --regions <FILE>
                        regions BED file in ATaRVa format.

Optional arguments:
  --non-ref             If set, will not consider "same as reference alleles" in the analysis. [default: All-loci-mode]
  -o <FILE>, --out <FILE>
                        output file name.
  --contigs CONTIGS [CONTIGS ...]
                        contigs to calculate mendelian concordance eg [chr1 chr2 ch3]. If not mentioned every contigs in the region file will be considered.
```
**NOTE: All input files(VCF & BED) should be bgzipped and tabix indexed**

### Output header
The output has the following fields as their column name:
|     Field     |         Descritption          |
|---------------|-------------------------------|
| Common loci | Number of loci present in all three trio samples |
| Total alleles | Total alleles considered (2 × Common loci) |
| Mendelian alleles | Alleles matching exact Mendelian inheritance |
| Off by One | Alleles differing by 1 bp from expected inheritance |
| Off by 1motif | Alleles differing by one repeat motif |
| Non mendelian alleles | Alleles with mismatches larger than one motif |

## Examples
### Basic mode
```bash
python MIRCHI.py -k kid.vcf.gz -d dad.vcf.gz -m mom.vcf.gz -r TR_regions.bed.gz
```
### Non-ref mode
```bash
python MIRCHI.py -k kid.vcf.gz -d dad.vcf.gz -m mom.vcf.gz -r TR_regions.bed.gz --non_ref
```
## Customization
To calculate Mendelian concordance using VCF files generated by tools other than ATaRVa, you need to modify the `extract_alleles` function. This function should be updated to extract alleles based on their length and store them using a set operation within a dictionary. The dictionary should use the locus coordinate as the key. eg: *'chr1:10000-10467' : {500, 564}*